astropy/time/core.py

```diff
@@ -874,12 +874,43 @@ def shape(self, shape):
                     reshaped.append(val)
 
     def _shaped_like_input(self, value):
-        out = value
-        if value.dtype.kind == 'M':
-            return value[()]
-        if not self._time.jd1.shape and not np.ma.is_masked(value):
-            out = value[()] if value.dtype.fields else value.item()
-        return out
+        if not isinstance(self._time.jd1, np.ndarray):
+            raise TypeError(
+                f"Time objects should always contain arrays internally but "
+                f"this one has {self._time.jd1} of type "
+                f"{type(self._time.jd1)} and {self._time.jd2} of type "
+                f"{type(self._time.jd2)}.")
+        elif self._time.jd1.shape:
+            if isinstance(value, np.ndarray):
+                out = value
+            else:
+                raise TypeError(
+                    f"JD is an array ({self._time.jd1!r}) but value "
+                    f"is not ({value!r})")
+        else:
+            # zero-dimensional array
+            if (isinstance(value, np.ndarray)
+                    and not value.shape
+                    and not np.ma.is_masked(value)):
+                out = value[()]
+                #out = value[()] if value.dtype.fields else value.item()
+            else:
+                out = value
+        # De-numpify types for the convenience of callers
+        if isinstance(out, np.float64):
```

Comment:
There are generally no restrictions on the type of `value` produced by a time format. The existing Time API is that scalar values are always converted to a Python-native type where possible via `item()` and we should maintain that.  For test cases consider:
```
    @property
    def value(self):
        return self.jd1.astype(int)

    @property
    def value(self):
        out = self.jd1 + 1j * self.jd2
        return out

    @property
    def value(self):
        out = np.empty(shape=self.jd1.shape, dtype='S10')
        out[()] = b'junk'
        return out

    @property
    def value(self):
        out = np.empty(shape=self.jd1.shape, dtype='O')
        out[()] = {'a': 1}
        return out
```

Reply:
This is covered by `test_custom_format_can_return_any_scalar`.

Should this include datetime64? The pre-existing test `test_datetime64_no_format` demands that datetime64 be able emerge from a Time object, but `.item()` turns that into a python `datetime`. There's also record dtypes, which lost their field names and accessors when `.item()`ized. (Both of these lead to test failures, so I have special-cased them.)